File: iosApp/iosApp/Modules/DegenCult/CultProposal ✅/CultProposalWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol CultProposalWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: CultProposalWireframeContext
    ) -> CultProposalWireframe
}

final class DefaultCultProposalWireframeFactory: CultProposalWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: CultProposalWireframeContext
    ) -> CultProposalWireframe {
        DefaultCultProposalWireframe(
            parent,
            context: context
        )
    }
}

final class CultProposalWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CultProposalWireframeFactory in
           DefaultCultProposalWireframeFactory()
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/cultproposal/CultProposalWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.cultproposal

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframe
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframeContext
import smartadapter.internal.extension.name

interface CultProposalWireframeFactory {
    fun make(parent: Fragment?, context: CultProposalWireframeContext): CultProposalWireframe
}

class DefaultCultProposalWireframeFactory: CultProposalWireframeFactory {

    override fun make(
        parent: Fragment?, context: CultProposalWireframeContext
    ): CultProposalWireframe = DefaultCultProposalWireframe(
        parent?.let { WeakRef(it) },
        context
    )
}

class CultProposalWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CultProposalWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCultProposalWireframeFactory()
        }
    }
}